packages/backend/src/api/controllers/ForcedActionsController.ts

```diff
@@ -0,0 +1,147 @@
+
+import { UserService } from '../../core/UserService'
+import { AssetRepository } from '../../peripherals/database/AssetRepository'
+import { PreprocessedAssetHistoryRepository } from '../../peripherals/database/PreprocessedAssetHistoryRepository'
+import { ControllerResult } from './ControllerResult'
+
+export class ForcedActionsController {
+  constructor(
+    private readonly userService: UserService,
+    private readonly preprocessedAssetHistoryRepository: PreprocessedAssetHistoryRepository<
+      AssetHash | AssetId
+    >,
+    private readonly assetRepository: AssetRepository
+  ) {}
+
+  async getSpotForcedWithdrawalPage(
+    givenUser: Partial<UserDetails>,
+    vaultId: bigint
+  ): Promise<ControllerResult> {
+    const user = await this.userService.getUserDetails(givenUser)
+
+    if (!user) {
+      return { type: 'not found', content: 'User must be logged in' }
+    }
+
+    const assets =
+      await this.preprocessedAssetHistoryRepository.getCurrentByPositionOrVaultId(
+        vaultId
+      )
+
+    const asset = assets[0]
+    if (!asset) {
+      return { type: 'not found', content: 'Vault is empty' }
+    }
+
+    if (asset.starkKey != user.starkKey) {
+      return { type: 'not found', content: 'Vault does not belong to user' }
+    }
+
+    const assetDetails = await this.assetRepository.findDetailsByAssetHash(
+      asset.assetHashOrId as AssetHash
+    )
+
+    const content = renderNewSpotForcedWithdrawPage({
+      user,
+      starkExAddress: EthereumAddress.ZERO,
+      positionOrVaultId: vaultId,
+      starkKey: asset.starkKey,
+      asset: {
+        hashOrId: asset.assetHashOrId,
+        balance: asset.balance,
+        details: assetDetails,
+        priceUSDCents: 0n,
+      },
+    })
+
+    return { type: 'success', content }
+  }
+
+  async getPerpetualForcedWithdrawalPage(
+    givenUser: Partial<UserDetails>,
+    positionId: bigint,
+    assetId: AssetId
+  ): Promise<ControllerResult> {
+    const user = await this.userService.getUserDetails(givenUser)
+
+    if (!user) {
+      return { type: 'not found', content: 'User must be logged in' }
+    }
+
+    const assets =
+      await this.preprocessedAssetHistoryRepository.getCurrentByPositionOrVaultId(
+        positionId
+      )
+
+    const asset = assets.find((asset) => asset.assetHashOrId === assetId)
+    if (!asset) {
+      return {
+        type: 'not found',
+        content:
+          'Position is empty or does not contain asset with given assetId',
+      }
+    }
+
+    if (asset.starkKey != user.starkKey) {
+      return { type: 'not found', content: 'Vault does not belong to user' }
+    }
+
+    const content = renderNewSpotForcedWithdrawPage({
+      user,
+      starkExAddress: EthereumAddress.ZERO,
+      positionOrVaultId: positionId,
+      starkKey: asset.starkKey,
+      asset: {
+        hashOrId: asset.assetHashOrId,
+        balance: asset.balance,
+        priceUSDCents: 0n,
```

Comment:
Is this correct? It's a collateral, so maybe this field is not used. But I prefer to ask :)

Reply:
It is not I suppose. To be honest I do not know what should go in this place but I saw how you did the spot page and it was like that. I thought that maybe it is only temporary so decided to go with it 😅